### ISSUE TITLE
Can neovim-qt connect to nvim using stdio?

### ISSUE THREAD

**Issue by tarruda** [equalsraf/neovim-qt#50]
Hi

I'm planning to make some changes in the "builtin UI" infrastructure for the next month(after 0.1 release). My goal is to make every UI an external program that is spawned by nvim at startup so projects like neovim-qt are basically on the same layer as the builtin TUI.

The idea is to have nvim spawn the UI program and communicate to it via stdin/stdout, so the UI program doesn't have to worry about connecting to neovim.

This also lets one use the nvim binary to spawn any UI, all nvim command line options will be readily available to neovim-qt. For example, here's an idea of how neovim-qt could be used to edit files from stdin(the exact mechanism used to select which UI will be spawned is still going to be determined):

```
echo "file contents"  | nvim -u NONE --ui=neovim-qt -
```

I'm also preparing to implement nvim "remote mode" where nvim will be able to connect to other instances, so no UIs have to worry about connectivity beyond stdin/stdout.

Can you prepare neovim-qt to work in this scenario? I think this will only simplify things since it will no longer be necessary to worry about sockets or named pipes.

**Comment by equalsraf** [comment]
Hi! Got a PR going for this, seems to work with

    :rpcstart('nvim-qt', ['--embed'])

Is that what you had in mind?

> This also lets one use the nvim binary to spawn any UI, all nvim command line options will be readily available to neovim-qt.

Yes, also makes sense to avoid some of the issues we have seen: e.g. Neovim starts and blocks on some error waiting for a keypress, then the UI connects and nothing happens.

An issue that comes to mind is this - if Neovim fails to start (on an early error) or if it can't find or launch the UI process. it will be impossible to report errors to the user.

> mechanism used to select which UI will be spawned is still going to be determined):

It might be enough to have `nvim --ui` and pick an option from nvimrc with the UI binary path - this way Neovim doesn't need to know any specifics about the UI. In Vim there was a `-g` option to run the GUI instead of the terminal version, so we can even be backwards compatible and use that.

**Comment by tarruda** [comment]
> Is that what you had in mind?

Yes, but neovim will do it automatically at startup unless --embed or --headless is passed

> An issue that comes to mind is this - if Neovim fails to start (on an early error) or if it can't find or launch the UI process. it will be impossible to report errors to the user.

This shouldn't happen in standard situations, we'll always distribute and install an UI with neovim. In any case we can always print a warning to stdout and start in headless mode as a fallback.

> It might be enough to have nvim --ui and pick an option from nvimrc with the UI binary path - this way Neovim doesn't need to know any specifics about the UI. 

I'm not sure about using the nvimrc. I'd like the UI to process to be spawned before nvimrc is loaded so it can have information about the RGB flag, which would be useful for setting color schemes compatible with Vim and also has `has('gui')` working in the nvimrc. One way to do it is have nvim ask the UI via msgpack-rpc if it supports RGB and then automatically "attach it" with the correct settings.

> In Vim there was a -g option to run the GUI instead of the terminal version, so we can even be backwards compatible and use that.

True. I think we'll have default names for the UI binaries: nvim-tui(distributed with nvim) and nvim-gui which is searched when -g is passed. Other than that I think we have use a command-line option and/or environment variables to override the binary name. We can also do things like detect if there's a tty on stdin/stdout/stderr and if not try to spawn a GUI.

**Comment by equalsraf** [comment]
> I'm not sure about using the nvimrc. I'd like the UI to process to be spawned before nvimrc is loaded so it can have information about the RGB flag, which would be useful for setting color schemes compatible with Vim and also has has('gui').

on one hand you need to know if the GUI is running to process nvimrc correctly on the other some nvimrc commands are only useful before the gui is running (like set guifont). The `has('gui')`case is the easier one - if the `-g` flag was passed then, even if the UI has not connected yet. I guess I assumed all UIs would have RGB.

Also if the user has multiple commands in nvimrc that change the layout `:set columns` `:color ...` and they are executed after the UI has attached we can run into some weird behaviour, like the window changing size/colors as nvimrc is processed.

Another thing to remember here is that Vim did have .vimrc and .gvimrc to keep things separate, so maybe that is the way to solve this.

**Comment by suy** [comment]
What about commands like `:gui`? Wouldn't it be more in line with the possibilities of neovim to have an auto command to say that a GUI attached or detached? One might want to detach from the GUI to continue from a remote computer from SSH for example (or the other way around).

**Comment by equalsraf** [comment]
> What about commands like :gui? Wouldn't it be more in line with the possibilities of neovim to have an auto command to say that a GUI attached or detached?

In Vim there was `GUIEnter` and `GUIFailed` events.

When running from the TUI it is not clear to me what `:gui` would do, e.g. would the TUI shutdown.

> One might want to detach from the GUI to continue from a remote computer from SSH for example (or the other way around).

Fair point. It makes things a bit trickier.

1. In my mind it is hard to deal with cases where we are using the TUI and then switch to using a  GUI, because of has('gui_running') and colorchsemes i.e. one of the UIs might work poorly and reparsing nvimrc is not an option, not to mention plugins - I see this as corner case though.
2. Detaching and re-ataching over SSH will have to be done using sockets - it does raise an interesting point on how to detach from stdin and keeping Neovim running in the background.
3. Right now `has('gui_running')` is true IF there is a UI attached - but the wording in the docs is actually
    
    > gui_running		Vim is running in the GUI, or it will start soon.

    so it makes sense to be able to start Neovim with `gui_running` as true even if it is not attached using stdin e.g. `nvim -g=nolaunch`

**Comment by tarruda** [comment]
> on one hand you need to know if the GUI is running to process nvimrc correctly on the other some nvimrc commands are only useful before the gui is running (like set guifont). The has('gui')case is the easier one - if the -g flag was passed then, even if the UI has not connected yet. I guess I assumed all UIs would have RGB.

> Also if the user has multiple commands in nvimrc that change the layout :set columns :color ... and they are executed after the UI has attached we can run into some weird behaviour, like the window changing size/colors as nvimrc is processed.

I'm aware of the problem, I plan to split UI initialization in two parts:

1- nvim querying information such as RGB to the UI
2- nvim attaching the UI to start sending notifications

The nvimrc will be sourced between the two, which will allow for UI configuration before the ui is initialized. We can have nvim call a rpc method such as "ui_init" that will be used as a signal to initialize UI graphics(after nvimrc and all configuration is ready)

A related problem I have been facing is how to decouple UI-specific settings from the core. So far the best idea is to implement some mechanism allowing UIs to listen for changes in global variables, so commands like `let g:neovim_qt_font = 'Ubuntu Mono'` can be used to configure the UI(even if called after nvimrc). This would allow the UI initialization process work like this:

- nvim spawns the UI process
- UI process starts its event loop to communicate with nvim without initializing any graphics(displaying windows in case of GUI)
- nvim calls `ui_get_parameters`(get information such as the RGB flag and rows/columns from the UI)
- before responding to `ui_get_parameters`, the UI calls some rpc method to subscribe to changes in a namespace of global variables. For example, neovim-qt can call `vim_subscribe('changed:neovim_qt_*')` to receive notifications when a global variable starting with `neovim_qt_` is changed.
- the UI responds to `ui_get_parameters` which allows nvim to continue the initialization process(such as sourcing nvimrc)
- after nvimrc is sourced, the UI will have received any changes to configuration variables
- nvim calls `ui_init` to tell the UI that it may display graphics now
- nvim starts sending redraw notifications.

Note that this method for initializing "builtin UIs" is more complex than the simple `ui_attach` currently used, but it will allow UI initialization to be fully deterministic. The `ui_attach` method is still going to be used when attaching to headless instances though(I will explain it below)

> Another thing to remember here is that Vim did have .vimrc and .gvimrc to keep things separate, so maybe that is the way to solve this.

As I explained above, we need a new mechanism to decouple UI configuration from the core, but for now we'll keep everything in nvimrc.

> What about commands like :gui? Wouldn't it be more in line with the possibilities of neovim to have an auto command to say that a GUI attached or detached? 

`:gui` can be implemented by killing the current TUI process, forking into the background and spawning a new GUI process. We can have an auto command for UI attach/detach, but it is not relevant to this scenario.

>  One might want to detach from the GUI to continue from a remote computer from SSH for example (or the other way around).

As I explained in the initial post, it will no longer be necessary for UIs to worry about this, nvim will take care of attaching/detaching from remote sessions. The UI will always receive redraw events from stdin even if it is "connected" to a nvim instance running in another computer.

This will be possible because nvim is perfectly capable of being a client of itself and I have an experimental proof of concept patch: https://github.com/neovim/neovim/issues/3119#issuecomment-130906214 . This patch adds a `rpcconnect()` function which can be used to return a channel id from a remote nvim instance and will be used to implement some compatibility with Vim `--remote-*` command line options.

But I also want to extend it with a `rpcattachui(channel_id)` function which will tells the current nvim instance to call `ui_attach` on `channel_id` and forward all msgpack_rpc received from the channel to the attached UI. This will also put nvim in a new mode(possibly called **r**emote mode akin to **t**erminal mode) that will forward every key press(except a special sequence used to detach) to the remote nvim. Pressing a special key sequence(which can be configured with **r**[nore]map akin to **t**[nore]map) will make nvim return from `rpcattachui()`.

The reason to follow this approach is to allow greater customization: One can build and customize a remote session manager in pure vimscript, and even script remote sessions. And the best part is that it will always be available to the user no matter which UI is being used.(UIs just have to talk stdio)

> In my mind it is hard to deal with cases where we are using the TUI and then switch to using a GUI, because of has('gui_running') and colorchsemes i.e. one of the UIs might work poorly and reparsing nvimrc is not an option, not to mention plugins - I see this as corner case though.

Colorschemes designed specifically from Vim can have problems with multiple UIs, but keep in mind that nvim is fine with having both TUI and GUI configuration in the same highlight group, and it will send the correct information to each UI based on the RGB flag. In other words, `has('gui_running')` is there simply for compatibility with Vim, if you develop a colorscheme for Nvim it will handle multiple attached UIs just fine(no matter if they are RGB or not) so you don't need to check for `gui_running`.

> Detaching and re-ataching over SSH will have to be done using sockets - it does raise an interesting point on how to detach from stdin and keeping Neovim running in the background.

As explained, nvim will forward it to the UI stdin.

> Right now has('gui_running') is true IF there is a UI attached - but the wording in the docs is actually

Yes, this was done as a workaround to the problem I described at the start of the comment: Nvim needs to split the builtin UI initialization in two parts so when nvimrc is sourced it will have the correct information(Right now the UI is only attached after nvimrc)
